hazelcast/src/main/java/com/hazelcast/client/impl/connection/tcp/TcpClientConnectionManager.java

```diff
@@ -462,6 +467,15 @@ Connection connect(Object target, Function<Object, Connection> getOrConnectFunct
         } catch (ClientNotAllowedInClusterException e) {
             logger.warning("Exception during initial connection to " + target + ": " + e);
             throw e;
+        } catch (TargetDisconnectedException e) {
+            connectionProcessListener.remoteClosedConnection(addressProvider.apply(target));
+            return null;
+        } catch (HazelcastException e) {
+            logger.warning("Exception during initial connection to " + target + ": " + e);
+            if (e.getCause() instanceof IOException) {
+                connectionProcessListener.connectionAttemptFailed(addressProvider.apply(target));
+            }
+            return null;
         } catch (Exception e) {
```

Comment:
What about this case? Shouldn't we call `connectionAttemptFailed`? 

I feel like instead of introducing extra branches, we could try to evaluate the exception here and call the necessary listener methods

Reply:
We don't know the reason in this case. It wasn't hit in any of the scenarios targeted in the description of [MC-1332](https://hazelcast.atlassian.net/browse/MC-1332) so I think we don't need to cover it.

Reply:
That's a good idea. Pushed, thanks.